docs/spec.md

```diff
@@ -507,23 +565,35 @@ Performs element-wise addition of two tensors `lhs` and `rhs` and produces a
 * For integers: integer addition.
 * For floats: `addition` from IEEE-754.
 * For complex numbers: complex addition.
+* For quantized types:
+  * `float_result = (lhs - zero_point(lhs)) * scale(lhs) +
+    (rhs - zero_point(rhs)) * scale(rhs)`.
+  * `rounded_result = round_nearest_even(float_result / scale(result))`.
+  * `result = clamp(storage_min(result), rounded_result + zero_point(result), storage_max(result))`.
 
 #### Inputs
 
 | Label | Name  | Type   | Constraints |
 |-------|-------|--------|-------------|
-| (I1)  | `lhs` | tensor | (C1)        |
-| (I2)  | `rhs` | tensor | (C1)        |
+| (I1)  | `lhs` | tensor | (C1)-(C5)   |
+| (I2)  | `rhs` | tensor | (C1)-(C5)   |
 
 #### Outputs
 
 | Name     | Type   | Constraints |
 |----------|--------|-------------|
-| `result` | tensor | (C1)        |
+| `result` | tensor | (C1)-(C5)   |
 
 #### Constraints
 
-* (C1) `lhs`, `rhs` and `result` have the same type.
+* (C1) `shape(lhs) = shape(rhs) = shape(result)`.
+* If the operation doesn't use quantized types:
+  * (C2) `element_type(lhs) = element_type(rhs) = element_type(result)`.
+* If the operation uses quantized types:
+  * (C3) `element_type(lhs) = element_type(rhs) = element_type(result)`,
+    except for scales and zero points which may differ.
+  * (C4) `abs(storage_min(lhs)) != abs(storage_max(lhs))`.
```

Comment:
Why do we need this constraint?

I thought `!quant.uniform<i8<-127,127>:f32, scale:zp>` has been used by some models.

Reply:
The constraint, specifically for AddOp,  is inspired from the TFLite quantization spec for AddOp [ref](https://www.tensorflow.org/lite/performance/quantization_spec#int8_quantized_operator_specifications).

Please let us know more on the use-cases.

Reply:
May you quote the line you are referring to as well?  I am not sure which line translates to `abs(storage_min(lhs)) != abs(storage_max(lhs))`.

IIUC, the asymmetric vs. symmetric comparison in the link is only on the "ability to use whatever min/max convenient" vs. "keep both end symmetric".  Asymmetric doesn't translate to "input can't be symmetric".

Reply:
IMO, the quoted line only says that they choose `[-128, 127]` as storage type min/max instead of imposing `abs(storage_min(lhs)) != abs(storage_max(lhs))`.

Reply:
The spec has some general constraints for the components `storage_min` and `storage_max` of quantized element type like
```
* (C4) `storage_min` and `storage_max` have the following values:
  * If `is_signed(storage_type)` and `d = num_bits(storage_type)`:
    * `storage_min = -2^(d-1)`, `storage_max = 2^(d-1)-1` OR
    * `storage_min = -2^(d-1)+1`, `storage_max = 2^(d-1)-1`.
  * If `is_unsigned(storage_type)`:
    * `storage_min = 0`, `storage_max = 2^d - 1`.
```

and some AddOp specific constraints (under section `add`)
```
  * (C3) element_type(lhs) = element_type(rhs) = element_type(result), except for scales and zero points which may differ.
  * (C4) `abs(storage_min(lhs)) != abs(storage_max(lhs))`.
```

Together it suggests that the min/max storage type for both the `lhs` and `rhs` for AddOp is -128/127.

With the new formulation proposed in https://github.com/openxla/stablehlo/pull/1352#discussion_r1157942416
AddOp should specify specific values -128/127 for the `storage_min`/`storage_max`.

Reply:
The latest revision LGTM (w.r.t. this comment).  This can be resolved.